### PR TITLE
Load transformed bean in correct CL

### LINKED ISSUE #5015: Injecting from another module causes ClassCastException
**Describe the bug**
The class to be injected is located in Maven dependency. During the runtime, injection fails with java.lang.ClassCastException. Both main project and dependency have beans.xml. If the injected class is moved to the main project, injection works.

Verified in dev mode.

**Expected behavior**
Class should be injected without error.

**Actual behavior**

When the @Inject variable is used for the first time, Quarkus throws exception:

```
2019-10-30 08:26:52,612 INFO  [io.quarkus] (main) Installed features: [agroal, cdi, hibernate-orm, jdbc-h2, narayana-jta, resteasy, resteasy-jsonb, smallrye-openapi, swagger-ui]
2019-10-30 08:27:18,622 ERROR [io.qua.ver.htt.run.QuarkusErrorHandler] (vert.x-worker-thread-1) HTTP Request to /person failed, error id: 96789028-ac59-4737-82a2-032c65c14f36-1: org.jboss.resteasy.spi.UnhandledException: java.lang.RuntimeException: Error injecting link.zeljko.quarkus.model.PersonRepository link.zeljko.quarkus.service.PersonService.personRepository
...
Caused by: java.lang.ClassCastException: link.zeljko.quarkus.model.PersonRepository_ClientProxy cannot be cast to link.zeljko.quarkus.model.PersonRepository
	at link.zeljko.quarkus.service.PersonService_Bean.create(PersonService_Bean.zig:151)
	... 31 more
```

**To Reproduce**
1. Create two projects, the main project, and dependency with injectable class. Add beans.xml to both.
2. mvn install dependency
3. Add dependency to the main project
4. Inject class from the dependency
5. Run main project in the dev mode.
6. Make operation that accesses the injected variable.

**Configuration**
```properties
quarkus.datasource.url=jdbc:h2:file:./persons
quarkus.datasource.driver=org.h2.Driver
quarkus.hibernate-orm.database.generation=update
quarkus.http.port=8000
```

**Screenshots**
(If applicable, add screenshots to help explain your problem.)

**Environment (please complete the following information):**
 * Output of uname -a or ver: Microsoft Windows [Version 10.0.18362.418]
 * Output of java -version:
java version "1.8.0_221"
Java(TM) SE Runtime Environment (build 1.8.0_221-b11)
Java HotSpot(TM) 64-Bit Server VM (build 25.221-b11, mixed mode)
 * GraalVM version (if different from Java):
 * Quarkus version or git rev: 0.26.1

**Additional context**
[quarkus.zip](https://github.com/quarkusio/quarkus/files/3787476/quarkus.zip)

### PR DESCRIPTION
**Pull request by stuartwdouglas**
If a class is part of a transformed archive then it will
be loaded by the RuntimeClassLoader, which means that all
beans associated with this class must also be loaded by
the RuntimeClassLoader.

Fixes #5015

Note that this has been manually verified, but does not include a test, as a test would require multiple additional maven modules to be created.

### CODE REVIEW

File: extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ArcProcessor.java
Comment (mkouba) [inline extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ArcProcessor.java]: Hm, I think that we should rather try to enhance the application class predicate used internally by ArC: https://github.com/quarkusio/quarkus/blob/master/extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ArcProcessor.java#L113-L119

This way we wouldn't have to modify the `Resource` class etc. WDYT?

Comment (stuartwdouglas) [reply extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ArcProcessor.java]: We will have a circular dependency issue if we use a transformer to remove final methods

Comment (mkouba) [reply extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ArcProcessor.java]: I see. In that case, we should add a comment here so that we're not trying to reimplement it later ;-).

File: extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ArcProcessor.java
Comment (mkouba) [inline extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ArcProcessor.java]: `if (resource.getAssociatedClassName() != null && transformedArchiveBuildItem.getClasses().contains(resource.getAssociatedClassName().toString())` ?

File: core/deployment/src/main/java/io/quarkus/deployment/steps/ClassTransformingBuildStep.java
Comment (gsmet) [inline core/deployment/src/main/java/io/quarkus/deployment/steps/ClassTransformingBuildStep.java]: `Files.walk()` doesn't need to be in a try with resources?

I have a vague recollection that it needs to be.

**Review by gsmet — CHANGES_REQUESTED** [review]
Added a small comment that we need to clear before this going in.

**Comment by gsmet** [comment]
@stuartwdouglas I'll let you decide but my opinion is that it's a bit too risky to merge this one at the very last minute.

**Comment by stuartwdouglas** [comment]
yea, lets hold off on this one.

**Comment by gsmet** [comment]
@stuartwdouglas could you rebase it and check @mkouba 's comment?

**Comment by gsmet** [comment]
@stuartwdouglas I think this one is ready for a rebase and should be included soon.

**Comment by stuartwdouglas** [comment]
This will be superseded by the new class loading stuff, but we might as well get it in for now.

**Comment by stuartwdouglas** [comment]
I am going to close this, I am hoping to have a first version of the new class loader work ready this week, and merging this would interfere with it.

**Comment by robertjsaulnier** [comment]
Hi, wondering if the new class loader stuff fixed this issue. I'm still getting the problem when running with quarkus:dev

**Comment by stuartwdouglas** [comment]
Can you open a new issue with a reproducer?

**Pull request closed without merging**
